src/verifast1.ml

```diff
@@ -2785,6 +2841,96 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
+      [t] -> [t]
+    | [] -> failwith "calculating least upper bound of no type"
+    | tps -> 
+      let st_map = List.map (fun tp -> (tp, tp::super_types tp)) tps in
+      let filtered_candidate_set = (*intersection of all est sets *)
+        let intersection setA setB = List.filter (fun memA -> List.mem memA setB) setA in
+        let rec iter sets iset = match sets with
+          [] -> []
+        | [last] -> intersection iset last
+        | hd::tl -> iter tl (intersection iset hd)
+        in
+        iter (List.map snd st_map) (snd (List.hd st_map))
+      in
+      (* MEC = { V | V in EC, and for all W ≠ V in EC, it is NOT the case that W <: V } *)
+      let mec = filtered_candidate_set |> List.filter begin fun v -> (* V in EC *)
+        let children = filtered_candidate_set |> List.filter begin fun w ->  (* for every V, select all it's children in EC that are not V *)
+          if (w = v) then   
+            false
+          else
+            (is_subtype_of_ w v)
+        end
+        in (* if V has no children, it is the most specific candidate*)
+        children = []
+      end
+      in
+      mec 
+
+  let infer_type l unknownTypes knownTypes =
+    let rec inference_set set actual expected = 
+      match (actual, expected) with 
+        (hd::tl, hd0::tl0) -> begin match (hd,hd0) with
+        | (ObjType (t, targs), ObjType (t0, targs0)) when t != "null" && t0 != "null" ->
+          if proper_type hd && proper_type hd0 then
+            inference_set set tl tl0
+          else if List.length targs != List.length targs0 then
+            inference_error l ("Type mismatch during inference, amount of type arguments does not match. Unknown type: " ^ string_of_type hd ^ ". Matching type: " ^ string_of_type hd0 ^ ".")
+          else if t != t0 then
+            inference_error l ("Type mismatch during inference. Unknown type: " ^ string_of_type hd ^ ". Matching type: " ^ string_of_type hd0 ^ ".")
+          else
+            inference_set set (targs@tl) (targs0@tl0)
+        | (InferredRealType t, ArrayType _) ->
+          inference_error l ("No inference for arrays yet.")
+        | (InferredRealType t, t0) ->
+          if is_primitive_type t0 then
+            inference_error l ("Can't infer primitive types yet, no boxing supported.")
+          else
+            inference_set ((t,t0)::set) tl tl0
+        | (ArrayType t, ArrayType t0) ->
+          inference_set set (t::tl) (t0::tl0)
+        | _ -> inference_set set tl tl0
+        end
+      | _ -> set
+    in
+    let binds = inference_set [] unknownTypes knownTypes in
+    let rec infvars binds vs =
+      match binds with
+        (infvar,tp)::tl ->
+          if List.mem infvar vs then
+            infvars tl vs
+          else
+            infvars tl (infvar::vs)
+      | [] -> vs
+    in
+    let setPerInfVar = (infvars binds []) |> List.map begin fun infvar -> 
+      (infvar, List.map snd (List.filter (fun (n,tp) -> n = infvar) binds))
+    end  

```

Comment:
Instead of `List.map begin fun infvar -> ... end`, write `List.map $. fun infvar -> ...`

Reply:
I'm not sure how $. f x = f x applies here?